File: 760. Find Anagram Mappings/using_hashmap_1.py
"""
760. Find Anagram Mappings
Easy

You are given two integer arrays nums1 and nums2 where nums2 is an anagram of nums1. Both arrays may contain duplicates.

Return an index mapping array mapping from nums1 to nums2 where mapping[i] = j means the ith element in nums1 appears in nums2 at index j. If there are multiple answers, return any of them.

An array a is an anagram of an array b means b is made by randomizing the order of the elements in a.

Example 1:

Input: nums1 = [12,28,46,32,50], nums2 = [50,12,32,46,28]
Output: [1,4,3,2,0]
Explanation: As mapping[0] = 1 because the 0th element of nums1 appears at nums2[1], and mapping[1] = 4 because the 1st element of nums1 appears at nums2[4], and so on.
Example 2:

Input: nums1 = [84,46], nums2 = [84,46]
Output: [0,1]
 

Constraints:

1 <= nums1.length <= 100
nums2.length == nums1.length
0 <= nums1[i], nums2[i] <= 105
nums2 is an anagram of nums1.
"""
from typing import List
class Solution:
    def anagramMappings(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums2_map = {key: i for i, key in enumerate(nums2)}
        #print(nums2_map)
        output = []
        for num in nums1:
            #print(num, nums2_map[num])
            output.append(nums2_map[num])
        return output
    
# O(n) time 
# O(n) space

def main():
    nums1 = [12,28,46,32,50]
    nums2 = [50,12,32,46,28]
    s=Solution()
    print(s.anagramMappings(nums1, nums2))

if __name__ == '__main__':
    main()

File: 760. Find Anagram Mappings/using_hashmap_1.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> anagramMappings(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> nums2_map;
        for (int i = 0; i < nums2.size(); i++) {
            nums2_map[nums2[i]] = i;
        }
        vector<int> output;
        for (int num : nums1) {
            output.push_back(nums2_map[num]);
        }
        return output;
    }
};

int main() {
    Solution s;
    vector<int> nums1 = {2, 4, 6, 8};
    vector<int> nums2 = {8, 2, 6, 4};
    vector<int> output = s.anagramMappings(nums1, nums2);
    for (int num : output) {
        cout << num << " ";
    }
    cout << endl;
    return 0;
}